### PR TITLE
authz/github: Support authorization.internalRepos in provider

### PR DESCRIPTION
**Pull request by indradhanush**
First step on #34684. 

This PR adds a new configurable attribute for the GitHub Authorization provider: `internalRepos`. Setting it from the UI is possible but has no affect as this is not being consumed anywhere at the moment. As a result it should be safe to merge this PR.

**Adding authorization.internalRepos to the config**

authorization-internal-repos

**UI Warning with setting both authorization.groupsCacheTTL and authorization.internalRepos in the config**

authorization-groups-cache-and-internal-repos

## Test plan

Added tests.

### CODE REVIEW

File: schema/github.schema.json [hunk 1]
```diff
@@ -168,6 +168,11 @@
           "description": "Experimental: If set, configures hours cached permissions from teams and organizations should be kept for. Setting a negative value disables syncing from teams and organizations, and falls back to the default behaviour of syncing all permisisons directly from user-repository affiliations instead. [Learn more](https://docs.sourcegraph.com/admin/repo/permissions#teams-and-organizations-permissions-caching).",
           "type": "number",
           "default": 72
+        },
+        "internalRepos": {
```
Comment (bobheadxi) [inline schema/github.schema.json:172]:
```suggestion
        "internal": {
```

as per your proposal https://github.com/sourcegraph/sourcegraph/issues/34684 , and I think it makes sense with my other comment

File: schema/github.schema.json [hunk 2]
```diff
@@ -168,6 +168,11 @@
...
+          "description": "Experimental: If set, configures the cached permissions to use a pseudo-org internally to support visibility of internal repositories for all users that belong to an organization. Ignored if GroupsCacheTTL is a positive number.",
```
Comment (bobheadxi) [inline schema/github.schema.json:173]:
```suggestion
          "description": "Experimental: If true, treats all repositories in this code host connection as visible to any user that is a member of any organization in the GitHub Enterprise instance. Ignored if GroupsCacheTTL is enabled. Not supported on GitHub.com.",
```

File: schema/github.schema.json (hunk 1 above)
Comment (indradhanush) [reply schema/github.schema.json:172]: I initially went with internal but then renamed this to internalRepos as the former felt less intuitive when reading its usage. How do you feel about it?

File: schema/github.schema.json (hunk 2 above)
Comment (bobheadxi) [reply schema/github.schema.json:173]: We should avoid mentioning implementation details and focus on what the options do here

File: schema/github.schema.json (hunk 1 above)
Comment (bobheadxi) [reply schema/github.schema.json:172]: I think internal makes sense: "everything syncd from this code host has internal visibility"

File: enterprise/internal/authz/github/authz.go [hunk 3]
```diff
@@ -91,6 +91,23 @@ func NewAuthzProviders(
...
+		// block, it will be certain if p.gropusCache is non nil, depeneding on which we make the
+		// next check.
+		//
+		// c.Authorization.Internal and c.Authorization.GroupsCacheTTL are mutually exclusive.
+		if p.groupsCache != nil && p.internalRepos {
+			warnings = append(warnings,
+				fmt.Sprintf("GitHub config for %[1]s has both `authorization.groupsCacheTTL` and "+
+					"`authorization.internalRepos` enabled, but they are mutually exclusive. "+
+					"`authorization.internalRepos` will be disabled. Update the code host config "+
+					"to resolve this warning.",
```
Comment (bobheadxi) [inline enterprise/internal/authz/github/authz.go:104]: Is this really the case? My impression was that it would be fairly easy to add for both

Comment (unknwon) [reply enterprise/internal/authz/github/authz.go:104]: Any chance could we prevent such config from saving?

File: schema/github.schema.json (hunk 1 above)
Comment (unknwon) [reply schema/github.schema.json]: I'm a bit confused here, my understanding of "internal repos" and "making all synced repos are internally visible" are two different things. Which are we addressing? If I understand correctly from https://github.com/sourcegraph/sourcegraph/issues/34684, we are addressing the former? But this PR seems addressing the latter?

**Comment by indradhanush** [comment]
Closing this for now to clean up the clutter. Deescalated in priority among other things for the time being. Will come back to this in the future and pick it up from where I've left off.

**Pull request closed without merging**